Thermal.py

```diff
@@ -0,0 +1,1390 @@
+import os
+os.environ['XLA_PYTHON_CLIENT_PREALLOCATE'] = 'false'
+import numpy as np
+from jax import jit
+import jax.numpy as jnp
+from jax.config import config
+config.update("jax_enable_x64", True)
+import scipy
+from functools import partial
+import Lanczos_lax
+
+
+def Random_force_on_grid_indexing(Nx,Ny,Nz):
+    normal_indices = []
+    normal_conj_indices = []
+    nyquist_indices = []
+    for i in range(Nx):
+        for j in range(Ny):
+            for k in range(Nz):
+                
+                if (not(2*k >= Nz+1) and 
+                    not((k == 0) and (2*j >= Ny+1)) and 
+                    not((k == 0) and (j == 0) and (2*i >= Nx+1)) and 
+                    not((k == 0) and (j == 0) and (i == 0))):
+                    
+                    ii_nyquist = ( ( i == int(Nx/2) ) and ( int(Nx/2) == int((Nx+1)/2) ) )
+                    jj_nyquist = ( ( j == int(Ny/2) ) and ( int(Ny/2) == int((Ny+1)/2) ) )
+                    kk_nyquist = ( ( k == int(Nz/2) ) and ( int(Nz/2) == int((Nz+1)/2) ) )
+                    if (   (i==0 and jj_nyquist and k==0) 
+                        or (ii_nyquist and j==0 and k==0) 
+                        or (ii_nyquist and jj_nyquist and k==0)
+                        or (i==0 and j==0 and kk_nyquist)
+                        or (i ==0 and jj_nyquist and kk_nyquist)
+                        or (ii_nyquist and j==0 and kk_nyquist)
+                        or (ii_nyquist and jj_nyquist and kk_nyquist)):
+                        nyquist_indices.append([i,j,k])
+                    else:
+                        normal_indices.append([i,j,k])
+                        if(ii_nyquist or (i==0)):
+                            i_conj = i    
+                        else:
+                            i_conj = Nx - i
+                        if(jj_nyquist or (j==0)):
+                            j_conj = j    
+                        else:
+                            j_conj = Ny - j
+                        if(kk_nyquist or (k==0)):
+                            k_conj = k    
+                        else:
+                            k_conj = Nz - k
+                        normal_conj_indices.append([i_conj,j_conj,k_conj])
+    
+    normal_indices = jnp.array(normal_indices)
+    normal_conj_indices = jnp.array(normal_conj_indices)
+    nyquist_indices = jnp.array(nyquist_indices)
+    
+    normal_indices_x = normal_indices[:,0]
+    normal_indices_y = normal_indices[:,1]
+    normal_indices_z = normal_indices[:,2]
+    normal_conj_indices_x = normal_conj_indices[:,0]
+    normal_conj_indices_y = normal_conj_indices[:,1]
+    normal_conj_indices_z = normal_conj_indices[:,2]
+    nyquist_indices_x = nyquist_indices[:,0]
+    nyquist_indices_y = nyquist_indices[:,1]
+    nyquist_indices_z = nyquist_indices[:,2]
+
+    return normal_indices_x,normal_indices_y,normal_indices_z, normal_conj_indices_x,normal_conj_indices_y,normal_conj_indices_z, nyquist_indices_x,nyquist_indices_y,nyquist_indices_z
```

Comment:
Is there a reason to return individual values instead of vectors?

Reply:
This is still a tricky point, originally it returned directly a vector but it also gave issues when jit compiling. So I changed to individual values, and later discovered that the issue was probably caused by something else. So probably it can be changed back. I guess for now we can leave it as it is.